ducktape/services/service.py

```diff
@@ -12,36 +12,63 @@
-# configure each server, and bring up/tear down the service. They also expose
-# information about the service so that other services or test scripts can
-# easily be configured to work with them. Finally, they may be able to collect
-# and check logs/output from the service, which can be helpful in writing tests
-# or benchmarks.
-#
-# Services should generally be written to support an arbitrary number of nodes,
-# even if instances are independent of each other. They should be able to assume
-# that there won't be resource conflicts: the cluster tests are being run on
-# should be large enough to use one instance per service instance.
 
 from ducktape.command_line.config import ConsoleConfig
 
 
 class Service(object):
-    def __init__(self, service_context):
-        """
-        :type service_context: ServiceContext
-        """
-        self.num_nodes = service_context.num_nodes
-        self.cluster = service_context.cluster
-        self.logger = service_context.logger
+    """Service classes know how to deploy a service onto a set of nodes and then clean up after themselves.
 
-    def start(self):
-        """Start the service running in the background."""
+    They request the necessary resources from the cluster,
+    configure each node, and bring up/tear down the service.
+
+    They also expose
+    information about the service so that other services or test scripts can
+    easily be configured to work with them. Finally, they may be able to collect
+    and check logs/output from the service, which can be helpful in writing tests
+    or benchmarks.
+
+    Services should generally be written to support an arbitrary number of nodes,
+    even if instances are independent of each other. They should be able to assume
+    that there won't be resource conflicts: the cluster tests are being run on
+    should be large enough to use one instance per service instance.
+    """
+
+    def __init__(self, context, num_nodes):
+
+        self.num_nodes = num_nodes
+        self.cluster = context.cluster
+        self.logger = context.logger
+        self.context = context
+        self.allocated = False
+
+        # Every time a service instance is created, it registers itself with its
+        # context object. This makes it possible for external mechanisms to clean up
+        # after the service if something goes wrong.
+        if hasattr(self.context, "services"):
+            self.context.services.append(self)
+
+        # Provides a mechanism for locating and collecting log files produced by the service on its nodes.
+        # logs is a dict with entries that look like log_name: {"path": log_path, "collect_default": boolean}
+        #
+        # For example, zookeeper service might have self.logs like this:
+        # self.logs = {
+        #    "zk_log": {"path": "/mnt/zk.log",
+        #               "collect_default": True}
+        # }
+        self.logs = {}
```

Comment:
You might not want to always set this here. In a lot of cases, there will be a fixed set of logs for a given service, in which case it might make more sense to set it at the class level, e.g. something like

```
class Zookeeper(Service):
    logs = { "zk_log": {"path": "/mnt/zk.log", "collect_default": True}
```

You can make this work well while still providing a default here (rather than having to always check `hasattr(self, 'logs')`) using a pattern like this example:

```
>>> class Base(object):
...     def __init__(self):
...             if not hasattr(self, 'logs'):
...                     self.logs = ''
...
>>> class Foo(Base):
...     logs = 'foo.log'
...
>>> Base().logs
''
>>> Foo().logs
'foo.log'
```